spark-statsd-reporter/src/test/java/org/apache/spark/metrics/sink/statsd/StatsdReporterTest.java

```diff
@@ -174,19 +180,150 @@ public void testTimers() throws Exception {
             assertEquals(14, server.receivedMessages().size());
         }
     }
+    
+    @Test
+    public void testExcludeFilter() throws Exception {
+        final int testUdpPort = 4445;
+        try (DatagramTestServer server = DatagramTestServer.run(testUdpPort)) {
+        	ImmutableSet<String> excludes = ImmutableSet.copyOf("test-app-01.driver.TestGaugeInt".split(","));
+        	StatsdReporterFactory reporterFactory = new StatsdReporterFactory();
+        	reporterFactory.setFormatter(formatter);
+        	reporterFactory.setHost("localhost");
+        	reporterFactory.setPort(testUdpPort);
+        	reporterFactory.setExcludes(excludes);
+        	StatsdReporter reporter = reporterFactory.build(null);
+
+            SortedMap<String, Gauge> gauges = new TreeMap<>();
+            gauges.put("test-app-01.driver.TestGaugeInt", () -> 42);
+            gauges.put("test-app-01.driver.TestGaugeFloat", () -> 123.123f);
+
+            reporter.report(gauges, new TreeMap<>(), new TreeMap<>(), new TreeMap<>(), new TreeMap<>());
+
+            sleep(100); // Pausing to let DatagramTestServer collect all messages before the assertion
```

Comment:
send code is synchronous, is `sleep` really needed? Also, why specifically 100ms is good enough to receive bytes on the other side of the socket, is it because of UDP? 

It's been like before in the reporter tests, so I'm just wondering what was the specific reason for that.

Reply:
Send code is synchronous, but DatagramTestServer is running in a separate thread. The pause allows it to get all datagrams before assertions are made `assertTrue(server.receivedMessages()...`

I had tests fail without the pause. I wish I knew a better solution than this, e. g. send some message to mark the end of the test and wait for server thread to get it, but UDP datagrams can be recieved out of order even locally.

Another option would be to wait for a specific number of datagrams, but so far 100ms pause worked without issues.